service/header/core_listener.go

```diff
@@ -36,34 +36,63 @@ func NewCoreListener(bcast Broadcaster, fetcher *core.BlockFetcher, dag format.D
 }
 
 // Start kicks off the CoreListener listener loop.
-func (cl *CoreListener) Start(ctx context.Context) error {
+func (cl *CoreListener) Start(context.Context) error {
 	if cl.cancel != nil {
 		return fmt.Errorf("core-listener: already started")
 	}
 
-	sub, err := cl.fetcher.SubscribeNewBlockEvent(ctx)
-	if err != nil {
-		return err
-	}
-
 	ctx, cancel := context.WithCancel(context.Background())
-	go cl.listen(ctx, sub)
+
+	go cl.listen(ctx)
 	cl.cancel = cancel
 	return nil
 }
 
 // Stop stops the CoreListener listener loop.
-func (cl *CoreListener) Stop(ctx context.Context) error {
+func (cl *CoreListener) Stop(context.Context) error {
 	cl.cancel()
 	cl.cancel = nil
-	return cl.fetcher.UnsubscribeNewBlockEvent(ctx)
+	return nil
 }
 
 // listen kicks off a loop, listening for new block events from Core,
 // generating ExtendedHeaders and broadcasting them to the header-sub
 // gossipsub network.
-func (cl *CoreListener) listen(ctx context.Context, sub <-chan *types.Block) {
-	defer log.Info("core-listener: listening stopped")
+func (cl *CoreListener) listen(ctx context.Context) {
+	var sub <-chan *types.Block
+
+	defer func() {
+		// only attempt to unsubscribe from new block events if currently
+		// subscribed
+		if sub != nil {
+			err := cl.fetcher.UnsubscribeNewBlockEvent(ctx)
+			if err != nil {
+				log.Errorw("core-listener: failed to unsubscribe from new block events", "err", err)
+			}
+		}
+		log.Info("core-listener: listening stopped")
+	}()
```

Comment:
Why did you take it out of Stop?

Reply:
Putting it back.

Reply:
Resolved? https://github.com/celestiaorg/celestia-node/pull/507/commits/d56b23295452c69dd15d5b13cf563814f7543735